### PR TITLE
[WIP] JS refactoring with revealing module pattern

### PR DESCRIPTION
**Pull request by rugk**
This PR fixes https://github.com/PrivateBin/PrivateBin/issues/178
Also fixes https://github.com/PrivateBin/PrivateBin/issues/76, fixes https://github.com/PrivateBin/PrivateBin/issues/176, fixes https://github.com/PrivateBin/PrivateBin/issues/187

## Changes

* Revealing Module Pattern
* "state" concept abandoned, now each module handles it's state internally (that's more flexible)
* more modules…
* a bit more console logging…

## ToDo

* [x] remove proxy, minimize usage of `this`
* [x] refactor controller (e.g. with own module for en/decryption)
* [x] try to make more methods private
* [x] ~~use classes for hiding/showing/… groups of elements (mostly TopNav…)~~
* [x] re-add SRI hash in bootstrap template (only removed for better dev)
* [x] make "page" template work with the refactored code
* [x] wait until translations are loaded before translating messages
* [x] make password protected pastes work again
* [x] readd error messages when trying to send empty pastes or on errors when sending comments
* [x] fix password protected burn after reading paste getting deleted when wrong password is entered, only delete when properly decrypted

## Known issues
* clicking on "new" does not reset expiration/format/… values. (that's acceptable for now IMHO)

### CODE REVIEW

**Comment by rugk** [comment]
There are two "Add comment" buttons when adding a comment. (above & below)
Is this new/intended?

**Comment by elrido** [comment]
Yes, the two buttons are intentional. If you have a very long discussion (no pagination yet, a long outstanding missing feature) it avoids having to scroll to find the button.

Some general questions for my understanding:
* What is the benefit of replacing this with me?
* Why replace the assignment of the methods object properties inline to separate assignments? It seems to require more boilerplate and be less readable.

**Comment by rugk** [comment]
> What is the benefit of replacing this with me?

First I though what (😯), then I understood you mean the variables…

So basically `me` is always the same as it is declared. `this` may change. Also we don't need any proxies anymore…

> Why replace the assignment of the methods object properties inline to separate assignments?

It's the Revealing Module Pattern. See https://toddmotto.com/mastering-the-module-pattern/
[One advantage is](https://carldanley.com/js-revealing-module-pattern) that we e.g. can have private methods and variables.

File: js/privatebin.js [hunk 1]
```diff
@@ -827,15 +718,15 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
                 {
                     if (paste.attachment)
                     {
-                        var attachment = filter.decipher(key, password, paste.attachment);
+                        var attachment = cryptTooldecipher(key, password, paste.attachment);
```
Comment (elrido) [inline js/privatebin.js]: Is it `cryptTooldecipher` or `cryptTool`**.**`decipher`?

Comment (rugk) [inline js/privatebin.js]: Yeah, thanks.

**Comment by rugk** [comment]
@elrido I find sjcl's behaviour strange. Obviously we use the latest version (1.0.6), but the [documented](https://bitwiseshiftleft.github.io/sjcl/doc/symbols/sjcl.prng.html#isReady) `prng` does not work. The same methods however work with `random`, which is not documented. 

Do you have any explanation for this? Otherwise I'd ask the sjcl people.

**Comment by elrido** [comment]
Thats actually documented at the top of [that same page](https://bitwiseshiftleft.github.io/sjcl/doc/symbols/sjcl.prng.html); `sjcl.prng` is the constructor called internally, but you should:

> Use sjcl.random as a singleton for this class!

**Comment by rugk** [comment]
I am also more or less fixing https://github.com/PrivateBin/PrivateBin/issues/93… Needs testing though.

**Comment by elrido** [comment]
So I was able to fix most unit testing issues except the one in the baseUri method. The problem on the Js side are two things:
1. The caching of things like the baseUri, pasteId or pasteKey induces state that needs to be reset between tests, so I had to add a reset function just for that. I think we don't gain much speed as these methods aren't doing heavy computation and are only called 1 - 3 times in one request anyway. Removing the caching there would reduce code complexity and the need for a reset function.
2. Injecting window and document seems to only benefit us in regards of documentation, but induce another cached state that we would need to be able to reset for unit testing. Is there any benefit to it except for code readability? I removed it in the Model object to make the unit tests work (these objects get destroyed and regenerated from scratch in the tests, not something that would happen in the browser). If there is a good reason to keep these, then readd them but provide a reset function to reset them for unit testing purposes. Removing them everywere would otherwise benefit the code size.

Otherwise there is still the page template to adapt to the new logic. I changed an ID that the unit tests showed as changed, but otherwise I did not yet look deeper into the causes for it to fail.

Oh, and I removed the purposeless span elements. If they have no attributes they have no function and browsers may just optimize them away in the DOM tree.

File: js/privatebin.js [hunk 2]
```diff
@@ -3413,8 +3421,10 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
             TopNav.hideAllButtons();
             Alert.showLoading('Sending comment…', 0, 'cloud-upload');
 
-            // get data
-            var [plainText, nickname] = DiscussionViewer.getReplyData(),
+            // get data, note that "var [x, y] = " structures aren't supported in all JS environments
```
Comment (rugk) [inline js/privatebin.js]: They are supported [since](https://stackoverflow.com/questions/6983026/javascript-assign-array-values-to-multiple-variables?noredirect=1&lq=1)

    FireFox 2.0+
    IE 9
    Opera 11.50.

So if you really care to support these horribly ancient browsers, okay…
Otherwise the array assignment is likely more readable. Low-prio though…

**Comment by rugk** [comment]
As for the `<span>` elements. They had their reason, or better… multiple ones, and indeed I failed to explain their purpose properly, so here it goes:
1. The main reason is to separate the icon (of the status, loading, etc. message) from the text. The icon is the `.glyphicon` element inside the parent element and the text is the second span (I just always addressed it as `:last` – Ctrl+F through the code to see where it is used). You could also address it with a class or so, … does not matter.
2. What I want to prevent is this (screenshot from current HEAD - 131e08ca33f49f0cbbc435f099437b2d06e46af4):
spanexm
2.1. The icon is clashed right next to the text. I added some margin-right to `.glyphicon` to prevent this without the unreliable and ugly solution to add spaces, which was (inconsistently, i.e. not in all places) used before.
2.2. You see the font/font style, that's because of bootstraps' styling of glyphicons.
2.3. To simplify this: We **must not** have text inside the `.glyphicon` class. E.g. also because of `aria-hidden=true` - likely a nightmare for users with screen readers. However we **must have** the possibility to replace the text of an item without having to replace the icon.
3. Another example from decrypting a paste. As you see there the loading text could not be replaced and somehow the old default text is still shown:
loading

In general look at [this code part](https://github.com/PrivateBin/PrivateBin/blob/131e08ca33f49f0cbbc435f099437b2d06e46af4/js/privatebin.js#L1053-L1073), which handles 
all messages (status, error, loading, expiry, …). See how the icon and text are selected with the `:first` and `:last` selector. The JS needs to be able to handle them differently to change icons, change the text etc.
This hole code breaks when their is no `:last` element. You see the consequences in my list above…

So if you have a better way to separate icon and text here and can change this in the JS. Otherwise I'd say the solution with just selecting the first and last element of such a status message, is quite easy.

**Comment by rugk** [comment]
Now your original comment:
1. Yes, the purpose of this caching is not really performance. It is rather a bit more stability and code style. E.g. take the `pasteId`. We need to get it during decryption, but may be later when doing certain things (adding comments, etc.). Currently we don't need this caching…
However I already thought about changing the URL bar to the home page or so, hiding the long URL when a paste is shown. In this case the data you can get from the model, needs to be available even *after* this URL change happened.
And IMHO this additional reset function is a good thing. Unused functions may become used later and if we already have a way to modify a class/module that's great.
Actually you might call this reset function (and possibly rename it is clear it only resets the things related to the paste) when a new paste is created or the paste is cloned. Both are cases where all this data gets invalid.

2. The document/window injection is best-practise as I read, to prevent accidental modification. However, feel free to remove them (i.e. you have already removed them, so that's okay). In the unit tests we of course want to do that modification deliberately, so yeah… I only added them because of code style/best practise reasons. They don't matter…

File: js/privatebin.js (hunk 2 above)
Comment (elrido) [inline js/privatebin.js]: Its not supported in node JS. No node JS, no unit tests...

**Comment by elrido** [comment]
Regarding the spans: The IMHO correct way to have this separation is to inject any number of whitespace characters (Browsers will always treat them as a single whitespace character, hence why having linebreaks and spaces/tabs for indentation counts as a single whitespace). Empty spans without classes or other properties can get optimized away.

Here is a compromise proposal: Readd the spans, but mark them with a class [following the proposed bootstrap v4 semantics](https://v4-alpha.getbootstrap.com/utilities/spacing/) for a padding of 1em, eg. `pl-1`.

**Comment by rugk** [comment]
As for the whitespace chars the only issue would be that we would have to use whitespaces in the translation files, whcih is ugly.
You can of course change it to use the bootstrap spacing. If this prevents any optimization, do it.

**Comment by elrido** [comment]
Why would you need to have whitespaces in the translation files? Those would be static in the template and not part of the translated string. That was how it was done so far and never was a problem before. The display glitches were introduced by this pull request.

But as said, if you want to handle that in CSS, please do, just make it explicit by assigning a class for this to the span, then it is clear it has a meaning.

**Comment by rugk** [comment]
The issue is that the message text is *not* static, so when we change the text we need to prefix it with a space. That's why I introduced the spans to solve this problem.

But it does not make sense to discuss about a solution we do not want to use. So please change it (back) the way you explained in https://github.com/PrivateBin/PrivateBin/pull/180#issuecomment-282542117 as your "compromise proposal". It's a good suggestion, so we do not need to discuss other potential solutions.

Of course if you could get it work with spaces and everything works (including message text changes etc.) you could also do so, but I think this is far difficult and unnecessary work. Just use the CSS solution… :blush:

**Comment by elrido** [comment]
There, I added your spans again, with CSS classes. Note that it was working perfectly fine with the spaces before your refactoring which introduced the glitch.

Could you now please fix the page template so that we can finally merge this and get on with all the other work?

File: js/privatebin.js [hunk 3]
```diff
@@ -278,14 +276,19 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
...
+                loc.href.length - loc.search.length - loc.hash.length
+            );
 
             // if base uri contains query string (when no base tag is present),
             // it is unwanted
-            if (baseUri.indexOf('?')) {
+            var queryIndex = baseUri.indexOf('?');
+            if (queryIndex !== -1) {
                 // so we built our own baseuri
-                baseUri = window.location.origin + window.location.pathname;
```
Comment (rugk) [inline js/privatebin.js]: Why does not `window.location.origin + window.location.pathname` work? Or what's bad about it?

**Comment by elrido** [comment]
RE: revert of baseUri code - I did not get it to work reliably in the unit test using that solution, so I reverted it to the one that the unit tests proves to work. I did not spend mutch time investigating, it could be that JSdom does not emulate the relatively new origin property, but from the detected errors I figured that certain corner cases were not handled correctly. If you get the unit test to work with your version, then we can of course leave it.

**Comment by elrido** [comment]
Ok, I found the cause of the missing origin. The [WhatWG-Url library](https://github.com/jsdom/whatwg-url/blob/07d451456c16e291d2945be12deab2d3f50c8dfc/src/url-state-machine.js#L1212) is to blame. When using randomly generated schemas it won't return an origin most of the time. I changed the test to only check for the supported schemas and voila: much simpler code.

**Comment by elrido** [comment]
To whom it may concern: @rugk informed me, that he wont have time to conclude this, so I'm taking this over.

File: js/privatebin.js [hunk 4]
```diff
@@ -1069,11 +1070,13 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
...
             if (args !== null) {
-                // add jQuery object to it as first parameter
-                args.unshift($element.find(':last'));
-
-                // pass it to I18n
-                I18n._.apply(this, args);
+                // get last text node of element
+                var content = $element.contents();
+                if (content.length > 1) {
+                    content[content.length - 1].nodeValue = ' ' + I18n._(args);
```
Comment (rugk) [inline js/privatebin.js]: No man, don't introduce this space again. It just sucks… The space, I mean.

I changed `I18n._` in a way that it can lazy-load the missing translation of an element if the JSON file is not already loaded (which actually happened in my tests), that's why you **have to** to pass a jQuery element to it for full compatibility. (at least for items, which might be loaded at the start, and error/status messages certainly belong to them).

File: js/privatebin.js [hunk 5]
```diff
@@ -1131,6 +1134,21 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
...
+         * This automatically passes the text to I18n for translation.
+         *
+         * @name   Alert.showRemaining
+         * @function
+         * @param  {string|array} message     string, use an array for %s/%d options
+         */
+        me.showRemaining = function(message)
+        {
+            console.error('remaining message shown: ', message);
+            handleNotification(1, $remainingTime, message);
```
Comment (rugk) [inline js/privatebin.js]: Ah right, good idea. Seems I missed to use the notification handler for this message.

File: js/privatebin.js [hunk 6]
```diff
@@ -1230,18 +1248,7 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
             $errorMessage = $('#errormessage');
             $loadingIndicator = $('#loadingindicator');
             $statusMessage = $('#status');
-
-            // display status returned by php code, if any (e.g. paste was properly deleted)
-            var serverStatus = $statusMessage.text();
```
Comment (rugk) [inline js/privatebin.js]: Sometimes the server sends status/error messages in the file. If you remove this, they might not be displayed. (I did not extensively test it, I just adapted this behaviour from pre-refactoring)

File: tpl/bootstrap.php [hunk 7]
```diff
@@ -379,31 +379,46 @@
...
-					<span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span><span class="pl-1"><?php echo htmlspecialchars($NOTICE), PHP_EOL; ?></span>
+					<span class="glyphicon glyphicon-info-sign" aria-hidden="true"></span>
+					<?php echo htmlspecialchars($NOTICE), PHP_EOL; ?>
 				</div>
 <?php
 endif;
 ?>
 				<div id="remainingtime" role="alert" class="hidden alert alert-info">
-					<span class="glyphicon glyphicon-fire" aria-hidden="true"></span><span class="pl-1"></span>
+					<span class="glyphicon glyphicon-fire" aria-hidden="true"></span>
```
Comment (rugk) [inline tpl/bootstrap.php]: Re-add the spans please. (see other comment) It will break when the JSON is not loaded otherwise.

File: js/privatebin.js (hunk 6 above)
Comment (elrido) [inline js/privatebin.js]: In my tests the opposite was the case: Due to this all errors or status messages of the server resulted in exceptions. The template ensures these are not hidden and work even if there is an error in the JS.

File: js/privatebin.js (hunk 4 above)
Comment (elrido) [inline js/privatebin.js]: Thanks for the tip reg. how the lazy loading now is supposed to work. Fixed it for anon comments and vizhash titles.

Reg. spans: I had taken the effort to readd them and waited weeks for you to finally fix the page tpl. You told me you currently have no time, fair enough. So I'll take this over and fix it so it works again, just as before. Spaces aren't broken, they are a feature of HTML. This was how PB has always worked, so how could it suddenly be broken except due to your code change. Not to speak of all the other mess you did not clean up: Passwords only worked with attachments, unit tests for both PHP and JS failing, etc. No time to fix stuff, but time to argue on minutiae. :angry:

**Comment by elrido** [comment]
Ok, I'll have to handle the remaining points another time. I'll see to it, that we can this finally merge within a week.

File: js/privatebin.js (hunk 6 above)
Comment (rugk) [inline js/privatebin.js]: Okay, if it works this way, that's good, but the original code also included this code.

File: js/privatebin.js [hunk 8]
```diff
@@ -463,7 +463,13 @@ jQuery.PrivateBin = function($, sjcl, Base64, RawDeflate) {
 
             // if $element is given, apply text to element
             if ($element !== null) {
-                $element.text(output);
+                // get last text node of element
+                var content = $element.contents();
+                if (content.length > 1) {
+                    content[content.length - 1].nodeValue = ' ' + output;
```
Comment (rugk) [inline js/privatebin.js]: Hmm, good idea, but we need to check for false-positives.
